### PR TITLE
Enable resetting of batchnorm running moments and cumulative average

### PR DESCRIPTION
**Pull request by jma127**
Duplicate of #5766

### CODE REVIEW

File: torch/nn/modules/batchnorm.py [hunk 1]
```diff
@@ -25,15 +25,21 @@ def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=True,
         if self.track_running_stats:
             self.register_buffer('running_mean', torch.zeros(num_features))
             self.register_buffer('running_var', torch.ones(num_features))
+            self.register_buffer('num_batches_tracked', torch.LongTensor([0]))
```
Comment (colesbury) [inline torch/nn/modules/batchnorm.py]: Will this break backwards compatibility? I think you may not be able to load old state_dicts because they will be missing `num_batches_tracked`?

**Comment by jma127** [comment]
cc @soumith

File: torch/nn/modules/batchnorm.py (hunk 1 above)
Comment (soumith) [reply torch/nn/modules/batchnorm.py]: yes this will. Which is why the PR was merged and reverted.
I think we should have a mechanism to plug-in backwards-compatibility for load_state_dict (we dont at the moment, unlike for standard pickling via `__setstate__`

**Comment by bddppq** [comment]
@onnxbot test this please

**Review by ezyang — CHANGES_REQUESTED** [review]
Blocking merge of this until BC-breakage is resolved

**Comment by jma127** [comment]
Thanks so much -- would be great to have it in time for the next release :)

**Comment by jma127** [comment]
Can someone help me out with this onnx check?

**Comment by soumith** [comment]
@SsnL please have a look

**Comment by soumith** [comment]
(closed by mistake)

File: torch/nn/modules/batchnorm.py [hunk 2]
```diff
@@ -44,14 +50,42 @@ def _check_input_dim(self, input):
...
+        try:
+            version = state_dict._metadata[prefix[:-1]]['version']
+        except (AttributeError, KeyError):
+            version = None
+
+        if version is None and self.track_running_stats:
+            num_batches_tracked_key = prefix + 'num_batches_tracked'
+            if num_batches_tracked_key not in state_dict:
+                # Add the missing num_batches_tracked counter
+                state_dict[num_batches_tracked_key] = torch.LongTensor([0])
```
Comment (SsnL) [inline torch/nn/modules/batchnorm.py]: `torch.tensor(0, dtype=torch.long)`

File: torch/nn/modules/batchnorm.py (hunk 1 above)
Comment (SsnL) [reply torch/nn/modules/batchnorm.py]: Please use `torch.tensor(0, dtype=torch.long)`

File: torch/nn/modules/batchnorm.py [hunk 3]
```diff
@@ -44,14 +50,42 @@ def _check_input_dim(self, input):
     def forward(self, input):
         self._check_input_dim(input)
 
+        exponential_average_factor = 0.0
+
+        if self.training and self.track_running_stats:
+            self.num_batches_tracked += 1
+            if self.momentum is None:  # use cumulative moving average
+                exponential_average_factor = 1.0 / max(1, self.num_batches_tracked.item())
```
Comment (SsnL) [inline torch/nn/modules/batchnorm.py]: why is the max needed?

**Review by ssnl — COMMENTED** [review]
I'm not sure if the new buffer should be in onnx export or not. If it should be, then updating the onnx expected output would be an easy fix.

**Comment by SsnL** [comment]
The TestCaffe2Backend.test_resnet  fail is at least easy to try locally. Could you see if it works?

**Comment by jma127** [comment]
comments addressed. I'm not sure what the TestCaffe2Backend test is, as grepping for that string returns nothing.

**Comment by dzhulgakov** [comment]
@onnxbot retest this please

**Review by ezyang — APPROVED** [review]

**Comment by jma127** [comment]
Woohoo! Thanks to all!
